src/util/rocksdb/properties.rs

```diff
@@ -466,6 +479,251 @@ impl DecodeProperties for UserCollectedProperties {
+            for v in handles {
+                buf.encode_u64(v.size).unwrap();
+                buf.encode_u64(v.offset).unwrap();
+            }
+        }
+        buf
+    }
+
+    fn decode(mut buf: &[u8], size: usize) -> Result<IndexHandlesMap> {
+        let mut res = IndexHandlesMap::new(size);
+        while !buf.is_empty() {
+            let klen = number::decode_u64(&mut buf)?;
+            let mut k = vec![0; klen as usize];
+            buf.read_exact(&mut k)?;
+            let mut handles = Vec::with_capacity(size);
+            for _ in 0..size {
+                let mut v = IndexHandle::default();
+                v.size = number::decode_u64(&mut buf)?;
+                v.offset = number::decode_u64(&mut buf)?;
+                handles.push(v);
+            }
+            res.handles.insert(k, handles);
+        }
+        Ok(res)
+    }
+
+    fn get_approximate_distance_in_range(&self, idx: usize, start: &[u8], end: &[u8]) -> u64 {
+        assert!(start <= end);
+        assert!(idx < self.size);
+        if start == end {
+            return 0;
+        }
+        let range = self.handles.range::<[u8], _>((Unbounded, Included(start)));
+        let start_offset = match range.last() {
+            Some((_, v)) => v[idx].offset,
+            None => 0,
+        };
+        let mut range = self.handles.range::<[u8], _>((Included(end), Unbounded));
+        let end_offset = match range.next() {
+            Some((_, v)) => v[idx].offset,
+            None => self.handles.iter().last().map_or(0, |(_, v)| v[idx].offset),
+        };
+        if end_offset < start_offset {
+            panic!(
+                "start {:?} end {:?} start_offset {} end_offset {}",
+                start, end, start_offset, end_offset
+            );
+        }
+        end_offset - start_offset
+    }
+
+    pub fn smallest_key(&self) -> Option<Vec<u8>> {
+        self.handles.iter().next().map(|(key, _)| key.clone())
+    }
+
+    pub fn largest_key(&self) -> Option<Vec<u8>> {
+        self.handles.iter().last().map(|(key, _)| key.clone())
+    }
+
+    pub fn iter(&self, idx: usize) -> IndexHandlesIterator {
+        IndexHandlesIterator::new(self.handles.iter(), idx)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct RangeProperties {
+    pub total_size: u64,
+    pub total_keys: u64,
+    pub index_handles: IndexHandlesMap,
+}
+
+impl RangeProperties {
+    pub fn encode(&self) -> UserProperties {
+        let mut props = UserProperties::new();
+        props.encode_u64(PROP_TOTAL_SIZE, self.total_size);
+        props.encode_u64(PROP_TOTAL_KEYS, self.total_keys);
+        props.encode(PROP_RANGE_INDEX, self.index_handles.encode());
+        props
+    }
+
+    pub fn try_decode<T: DecodeProperties>(props: &T) -> Result<RangeProperties> {
+        match RangeProperties::decode(props) {
+            Ok(res)=>return Ok(res),
+            Err(e)=>warn!("old_version:can not decode to RangeProperties:{:?}, try to decode from SizeProperties",e),
```

Comment:
This will print a lot of warnings when we upgrade an existing cluster.
I think it is not necessary to add a `try_decode` here, we should hide the details.